paddle/fluid/operators/mkldnn/softmax_mkldnn_op.cc

```diff
@@ -34,47 +34,61 @@ using platform::to_void_cast;
 
 class SoftmaxMKLDNNHandler : public platform::MKLDNNHandler {
  public:
-  SoftmaxMKLDNNHandler(const platform::MKLDNNDeviceContext& dev_ctx,
+  SoftmaxMKLDNNHandler(const std::vector<int>& dims,
+                       framework::proto::VarType::Type vtype,
+                       const mkldnn::memory::data_type dtype,
+                       const mkldnn::memory::format fmt,
+                       const platform::MKLDNNDeviceContext& dev_ctx,
+                       mkldnn::engine engine, const std::string& base_key)
+      : platform::MKLDNNHandler(dev_ctx, engine, base_key),
+        dims_(dims),
+        vtype_(vtype),
+        dtype_(dtype),
+        fmt_(fmt) {}
+
+  SoftmaxMKLDNNHandler(const std::vector<int>& dims,
+                       framework::proto::VarType::Type vtype,
+                       const mkldnn::memory::data_type dtype,
+                       const mkldnn::memory::format fmt,
+                       const mkldnn::memory::format diff_fmt,
+                       const platform::MKLDNNDeviceContext& dev_ctx,
                        mkldnn::engine engine, const std::string& base_key)
-      : platform::MKLDNNHandler(dev_ctx, engine, base_key) {}
-
-  SoftmaxMKLDNNHandler(
-      std::shared_ptr<mkldnn::softmax_forward::primitive_desc> softmax_pd,
-      std::shared_ptr<mkldnn::softmax_backward::primitive_desc> softmax_bwd_pd,
-      const platform::MKLDNNDeviceContext& dev_ctx, mkldnn::engine engine,
-      const std::string& base_key)
       : platform::MKLDNNHandler(dev_ctx, engine, base_key),
-        softmax_pd_(softmax_pd),
-        softmax_bwd_pd_(softmax_bwd_pd) {
+        dims_(dims),
+        vtype_(vtype),
+        dtype_(dtype),
+        fmt_(fmt),
+        diff_fmt_(diff_fmt) {
     // If we are in Grad operatgor then update a key with BWD suffix to
     // distinguish from FWD memory primitives
+    // Key_common will allow to access FWD_PD from cache
     key_ += "-BWD";
   }
 
-  std::shared_ptr<softmax_forward::primitive_desc>
-  AcquireSoftmaxPrimitiveDescriptor(const softmax_forward::desc& softmax_desc,
-                                    const mkldnn::engine& engine) {
-    // Softmax PD has to be passed to Grad op that
-    // may be executed by diffrent thread, hence
-    // for that one we use key that does not contain TID
-    const std::string key_softmax_pd = key_common_ + "@softmax_pd";
+  // TODO(jczaja): Once fwd_pd_ are moved to MKLDNNHandler then this function
+  // should be moved as well eg. SoftmaxMKLDNNHandler -> MKLDNNHandler<softmax_>
+  std::shared_ptr<mkldnn::memory> AcquireSrcMemory(void* ptr) {
+    return this->AcquireMemory(dims_, dtype_, fmt_, ptr, "@user_src_mem_p");
+  }
```

Comment:
I think we should get memory descriptors from mkldnn::query functions invoked on primitive descriptions instead of creating new memory descriptors with dims_, dtype_ and fmt_.

Reply:
In this code we create Memory object for Tensor data as it comes to MKL-DNN op. This not necessery is the same layout/format as indicated by computational primitive 's PD .

Reply:
I understand that user_src_md is not always the same as src_md but in this case you create softmax primitive descriptor this way:
`auto md = mkldnn::memory::desc(dims_, dtype_, fmt_);
auto softmax_desc = softmax_forward::desc(prop_kind::forward_scoring, md, 1 /*dim: C*/);` 
So mkldnn::query such as `fwd_pd_->src_desc()` = `mkldnn::memory::desc(dims_, dtype_, fmt_)`